### ISSUE TITLE
Cookbook Cleanup

### ISSUE THREAD

**Issue by RyanJarv** [rackspace-cookbooks/varnish#113]
I was going through this cookbook to add support for varnish 4.1 and ended up rewriting a pretty good chunk of it before trimming it down and submitting this pull request https://github.com/rackspace-cookbooks/varnish/pull/111#issuecomment-262330481.

Wanted to mention a few of the opinions/thoughts I had though and see if you would be open to accepting a partial rewrite.

- The default recipe should be using the lwrp's, maintaining two seperate versions of the templates is tedious.
- Generic vcl_file and vcl_template lwrp's would be more useful then varnish_default_vcl
- For users using resources installing varnish is simple and probably better left up to them. I think having a varnish repo lwrp makes sense though.
- The user should be able to override the default varnish lwrp template.
- Thinking it would be better to use node attributes to determine correct version and platform defaults for lwrps to avoid duplication.
- If breaking backwards compatibility isn't an issue would be nice to use the newer style resources.

There's a couple other things that would probably be good to address during a rewrite as well like the CentOS user/group settings and integration tests not working, varnish 5 support, HTTPS for apt public key, vmods (although I personally use packages.. so need to think if I should take that on or not). 

 Let me know if you have any thoughts on any of this.

**Comment by martinb3** [comment]
Hi there -- I'd be glad to review re-writes that do these things. I definitely agree that we should be using the LWRPs in the recipes. That was the intended thing all along, but the original author never finished it. I'd also be interested in doing a major release to make some of these breaking changes (repo LWRP, overriding templates, etc).

**Comment by RyanJarv** [comment]
Hey, great to hear. It would be awesome to see this cookbook be easier to contribute to and more reusable. 

I have been thinking about the usefulness of the default recipe. It's tempting to have a recipe that you all you just need to include to get varnish up and running, but I'm thinking this is just hiding some of the complexity of varnish.

Varnish may work with the default recipe provided but 99% of the time people are going to want to do more then that.. which granted is why node['varnish']['vcl_source'] and node['varnish']['vcl_cookbook'] and all the other attributes exist, but when we do stuff like that it add's complexity to the cookbook and we need to either decide between supporting those same attributes in the resources (global state for what most people see as functions) or have a large list of attributes that are only supported in the recipe and rely on the user to figure out what is supported in each case by reading the documentation.

I think for the purposes of creating a minimal config to get varnish up and running it might be better just to suggest users to look at the install_varnish fixture cookbook for copy/paste examples. This would also leave the default recipe open for various setup, workarounds, and hacks might be needed to use the resources, things like the reload-varnish script not supporting 4.1 in the upstream packaging.

Let me know if you have any thoughts on this.

**Comment by RyanJarv** [comment]
Here is the current rework (https://github.com/RyanJarv/varnish/tree/rewrite). I opted to remove the default recipe in favor of users just using the resources. Examples of what a recipe might look like can be found at (https://github.com/RyanJarv/varnish/blob/rewrite/test/fixtures/cookbooks/install_varnish/recipes/vendor_install.rb).

All the same parameters are accessible through the resources so the default recipe could easily be added back if we want to do that. Right now the default recipe just contains some of the workarounds that might be needed to use the resources (and that should be fixed upstream).

Messing around with it a bit it seems shelling out is probably the best way to determine the version so I left that in. I don't think there is really any issue with this, and made sure you can override it via resources properties incase you want to have multiple versions running.

Moving to the newer style resources saved a decent amount of lines and avoided some of the hacks that where happening before for triggering resources.

I'll still be working on this so right now I'm just posting it for anyone to review. Suggestions and comments are appreciated.

**Comment by martinb3** [comment]
Hey! 

> I have been thinking about the usefulness of the default recipe. It's tempting to have a recipe that you all you just need to include to get varnish up and running, but I'm thinking this is just hiding some of the complexity of varnish.

TBH, when I've done this in other projects, we inevitably get a lot of Github issues popped complaining. It's hard to judge how much the recipes are used. I ended up doing something like this in the elasticsearch one, which allows you to set any parameter on the resource using the node attributes, but doesn't require us to manage them all:
```
elasticsearch_user 'elasticsearch' do
  node['elasticsearch']['user'].each do |key, value|
    # Skip nils, use false if you want to disable something.
    send(key, value) unless value.nil?
  end
end
```

I totally agree that this adds complexity, even the way I posted above. We probably have to be prepared for a lot of issues opened either way 👍 

Carry on!

**Comment by martinb3** [comment]
PS -- just reviewed the diff & tests you've added... it looks good! :)

**Comment by RyanJarv** [comment]
Ok, I think that would work pretty well. I'm wondering if this will be too much effort for users to to upgrade now though.. We could maybe just mark it as deprecated and allow it to wrap another recipe that we will want to use long term, then when it's removed it would leave the default.rb recipe open for more generic stuff. Should be easy to do the same with the varnish_install and varnish_default_vcl resources as well.

I should mention the newer style resources only work on chef client 12.5 and newer which was released about a year ago now. I've seen a few recipes starting to use them now so I think It's probably ok but not sure what you are aiming to support, It would definitely rule out Chef 11 compatibility (https://github.com/rackspace-cookbooks/varnish/issues/82).

**Comment by martinb3** [comment]
Yeah -- I think both of the considerations you mention would probably mean we should bump major versions, and break Chef 11 compatibility as well as previous cookbook versions. I think that's okay 👍

**Comment by martinb3** [comment]
Merged #115. Going to close this, but I think we could probably close this and open any remaining individual issues we need to fix. Thanks again @RyanJarv!
